### PR TITLE
NDRS-325 put deploy variants

### PR DESCRIPTION
**Pull request by dwerner**
Here's my recent work on the options in NDRS-325

### CODE REVIEW

File: client/src/deploy/creation_common.rs [hunk 1]
```diff
@@ -777,6 +866,24 @@ pub(super) fn apply_common_session_options<'a, 'b>(subcommand: App<'a, 'b>) -> A
...
         )
+        .arg(session_package_hash::arg())
+        .arg(session_package_name::arg())
+        .arg(session_hash::arg())
+        .arg(session_name::arg())
+        .arg(session_entry_point::arg())
+        .arg(session_version::arg())
+}
+
+pub(crate) fn apply_common_purchase_options(
```
Comment (EdHastingsCasperLabs) [inline client/src/deploy/creation_common.rs:877]: where is this term `purchase` coming from?

**Review by EdHastingsCasperLabs — APPROVED** [review]
Optimistic approval assuming purchase / payment mix up is fixed.

File: client/src/deploy/creation_common.rs (hunk 1 above)
Comment (dwerner) [reply client/src/deploy/creation_common.rs]: yeah that should be payment

File: client/src/deploy/creation_common.rs [hunk 2]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+            .display_order(DisplayOrder::SessionHash as usize)
+    }
+
+    pub fn get(matches: &ArgMatches) -> Option<ContractHash> {
+        get_contract_hash(ARG_NAME, matches)
+    }
+}
+
+pub(crate) mod session_name {
+    use super::*;
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1050]:
```suggestion
    use super::*;

```

File: client/src/deploy/creation_common.rs [hunk 3]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+    pub const ARG_NAME: &str = "session-name";
+    const ARG_VALUE_NAME: &str = "NAME";
+    const ARG_HELP: &str = "Name of the stored contract (associated with the executing account) to be called as the session";
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1053]:
```suggestion
    const ARG_HELP: &str = "Name of the stored contract (associated with the executing account) to be called as the session";

```

File: client/src/deploy/creation_common.rs [hunk 4]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+    pub fn arg() -> Arg<'static, 'static> {
+        Arg::with_name(ARG_NAME)
+            .long(ARG_NAME)
+            .value_name(ARG_VALUE_NAME)
+            .help(ARG_HELP)
+            .required(false)
+            .requires(session_entry_point::ARG_NAME)
+            .display_order(DisplayOrder::SessionName as usize)
+    }
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1062]:
```suggestion
    }

```

File: client/src/deploy/creation_common.rs [hunk 5]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+    pub fn get(matches: &ArgMatches) -> Option<String> {
+        matches.value_of(ARG_NAME).map(str::to_string)
+    }
+}
+
+pub(crate) mod session_package_hash {
+    use super::*;
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1069]:
```suggestion
    use super::*;

```

File: client/src/deploy/creation_common.rs [hunk 6]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+            .display_order(DisplayOrder::SessionPackageHash as usize)
+    }
+
+    pub fn get(matches: &ArgMatches) -> Option<ContractHash> {
+        get_contract_hash(ARG_NAME, matches)
+    }
+}
+
+pub(crate) mod session_package_name {
+    use super::*;
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1090]:
```suggestion
    use super::*;

```

File: client/src/deploy/creation_common.rs [hunk 7]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+            .display_order(DisplayOrder::SessionPackageName as usize)
+    }
+
+    pub fn get(matches: &ArgMatches) -> Option<String> {
+        matches.value_of(ARG_NAME).map(str::to_string)
+    }
+}
+
+pub(crate) mod session_entry_point {
+    use super::*;
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1111]:
```suggestion
    use super::*;

```

File: client/src/deploy/creation_common.rs [hunk 8]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+            .display_order(DisplayOrder::SessionEntryPoint as usize)
+    }
+
+    pub fn get(matches: &ArgMatches) -> Option<String> {
+        matches.value_of(ARG_NAME).map(str::to_string)
+    }
+}
+
+pub(crate) mod session_version {
+    use super::*;
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1131]:
```suggestion
    use super::*;

```

File: client/src/deploy/creation_common.rs [hunk 9]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+    pub const ARG_NAME: &str = "session-version";
+    const ARG_VALUE_NAME: &str = "NAME";
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1133]:
```suggestion
    const ARG_VALUE_NAME: &str = "INTEGER";
```

File: client/src/deploy/creation_common.rs [hunk 10]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+    pub fn get(matches: &ArgMatches) -> Option<u32> {
+        matches
+            .value_of(ARG_NAME)
+            .map(|s| s.parse::<u32>().ok())
+            .flatten()
+    }
+}
+
+pub(crate) mod payment_hash {
+    use super::*;
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1154]:
```suggestion
    use super::*;

```

File: client/src/deploy/creation_common.rs [hunk 11]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+            .display_order(DisplayOrder::PaymentHash as usize)
+    }
+
+    pub fn get(matches: &ArgMatches) -> Option<ContractHash> {
+        get_contract_hash(ARG_NAME, matches)
+    }
+}
+
+pub(crate) mod payment_name {
+    use super::*;
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1175]:
```suggestion
    use super::*;

```

File: client/src/deploy/creation_common.rs [hunk 12]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+            .display_order(DisplayOrder::PaymentName as usize)
+    }
+
+    pub fn get(matches: &ArgMatches) -> Option<String> {
+        matches.value_of(ARG_NAME).map(str::to_string)
+    }
+}
+
+pub(crate) mod payment_package_hash {
+    use super::*;
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1197]:
```suggestion
    use super::*;

```

File: client/src/deploy/creation_common.rs [hunk 13]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+            .display_order(DisplayOrder::PaymentPackageHash as usize)
+    }
+
+    pub fn get(matches: &ArgMatches) -> Option<ContractHash> {
+        get_contract_hash(ARG_NAME, matches)
+    }
+}
+
+pub(crate) mod payment_package_name {
+    use super::*;
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1218]:
```suggestion
    use super::*;

```

File: client/src/deploy/creation_common.rs [hunk 14]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+            .display_order(DisplayOrder::PaymentPackageName as usize)
+    }
+
+    pub fn get(matches: &ArgMatches) -> Option<String> {
+        matches.value_of(ARG_NAME).map(str::to_string)
+    }
+}
+
+pub(crate) mod payment_entry_point {
+    use super::*;
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1239]:
```suggestion
    use super::*;

```

File: client/src/deploy/creation_common.rs [hunk 15]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+            .display_order(DisplayOrder::PaymentEntryPoint as usize)
+    }
+
+    pub fn get(matches: &ArgMatches) -> Option<String> {
+        matches.value_of(ARG_NAME).map(str::to_string)
+    }
+}
+
+pub(crate) mod payment_version {
+    use super::*;
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1259]:
```suggestion
    use super::*;

```

File: client/src/deploy/creation_common.rs [hunk 16]
```diff
@@ -907,3 +1014,266 @@ pub(super) mod input {
...
+    pub const ARG_NAME: &str = "payment-version";
+    const ARG_VALUE_NAME: &str = "NAME";
```
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:1261]:
```suggestion
    const ARG_VALUE_NAME: &str = "INTEGER";
```

**Review by Fraser999 — CHANGES_REQUESTED** [review]
Looking good.  Mainly nitpicks with the exception that `apply_common_payment_options` should be applied to the `make-deploy` subcommand too.

**Comment by dwerner** [comment]
@Fraser999 I tried to address your comments in my last commit. Please have another review when you get a chance.

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**